File: smama.pl
#!/usr/bin/env perl
# echo 今日は良い天気です。エロいピエロです。 | ./smama.pl smama-dic.tsv
use strict;
use warnings;
use utf8;
use open ':utf8';
binmode STDIN, ":utf8";
binmode STDOUT, ":utf8";

### (B1) 辞書ファイルを読み込む
# FORMAT: ^エントリ文字列\tマッチした時に表示する情報
my $dic_filename = shift @ARGV;
open(my $fh, "<", $dic_filename) or die;
my %dic = map {/^(.+?)\t(.+)$/; $1 => $2} <$fh>;
close($fh);

### (B2) エントリ文字列を繋げて正規表現用のパターン文字列を作る
my $pat = join("|", map {"\Q$_\E"} sort {length($b) <=> length($a)} keys %dic);

while (my $sentence = <>) {
    chomp $sentence;
    print "入力文: $sentence\n";

    ### (M1) 文を形態素に分割
    my $tokens = wakati($sentence);
    print "形態素: ".join("", map {"[$_]"} @$tokens)."\n";

    ### (M2) 形態素区切り位置を保存
    my $i = 0;
    my %kugiri = map {($i += length($_)) => 1} "", @$tokens;    
    print "区切り位置: ".join(" ", sort {$a <=> $b} keys %kugiri)."\n\n";

    ### (M3) 辞書にマッチさせる
    # 正規表現でマッチしたエントリ文字列のマッチ開始・終了位置の両方が
    # 形態素区切り位置に一致していればマッチ成功
    while ($sentence =~ /($pat)/g) {
	my ($from, $to) = ($-[0], $-[0] + length($1));
	if ($kugiri{$from} and $kugiri{$to}) {
	    print "MATCH! $from-$to [$1][$dic{$1}]\n";
	}
    }
}

### 形態素解析WebAPIで文を形態素に分割する
sub wakati {
    use LWP::UserAgent;
    use JSON;
    my ($query) = @_;
    my $appid = "THISISAPEN"; # 自分で取得した AppID を使いましょう
    my $ua = LWP::UserAgent->new;
    $ua->default_header('Content-Type' => 'application/json');
    $ua->default_header('User-Agent' => 'Yahoo AppID: '.$appid);
    my $url = "https://jlp.yahooapis.jp/MAService/V2/parse";
    my $params = {
        "jsonrpc" => "2.0",
        "method" => "jlp.maservice.parse",
	"id" => "123",
        "params" => { "q" => $query }
    };
    my $response = $ua->post($url, content => JSON::encode_json($params));
    if ($response->is_success) {
	my $j = decode_json($response->content);
	return [map {$_->[0]} @{$j->{"result"}{"tokens"}}];
    } else {
        die $response->status_line;
    }
}

File: smama.py
### 形態素解析WebAPI結果から形態素の表層文字列をリストとして返す関数
def sentence_to_tokens(query):
    import json
    from urllib import request
    APPID = "THISISAPEN"
    URL = "https://jlp.yahooapis.jp/MAService/V2/parse"
    params = json.dumps({
        "id": "1234-1",
        "jsonrpc": "2.0",
        "method": "jlp.maservice.parse",
        "params": { "q": query }
    }).encode()
    headers = {
        "Content-Type": "application/json",
        "User-Agent": "Yahoo AppID: {}".format(APPID),
    }
    req = request.Request(URL, params, headers)
    with request.urlopen(req) as res:
        body = res.read()
    obj = json.loads(body.decode())
    return [x[0] for x in obj['result']['tokens']]

### 入力文とマッチさせる辞書
sentence = 'ピエトロ氏、ドレッシングはピエロ事件。エロサイト'
dic = {
    'エロ': 'ero',
    'シング': 'sing',
    'ピエトロ': 'pietro',
    '良い天気': 'yoitenki',
    '事件': 'jiken',
}
print('入力文:', sentence)

### 各エントリが一箇所でもマッチしてればOK (マッチ位置や複数マッチは非対応)
# for entry in dic.keys():
#    if entry in sentence:
#        print("MATCH!", entry, dic[entry])

### (C1) 正規表現でマッチさせる
import re
#pattern = '|'.join(map(lambda x: re.escape(x), dic.keys()))
pattern = '|'.join(map(lambda x: re.escape(x),
                       sorted(dic.keys(), key=len, reverse=True)))

print(pattern)
for m in re.finditer(pattern, sentence):
    print("MATCH!", m.span(), m.string[m.start():m.end()])

### (M1) 文を形態素に分割
tokens = sentence_to_tokens(sentence)
print('形態素:', tokens)

### (M2) 形態素区切り位置を保存
kugiri_positions = set(sum(len(token) for token in tokens[:i])
                       for i in range(0, len(tokens) + 1))
print('区切り位置:', kugiri_positions, '\n')

### (M3) 辞書にマッチさせる
for current_position in kugiri_positions:
    for entry in dic.keys():
        end_position = current_position + len(entry)
        if (end_position in kugiri_positions) & \
           (sentence[current_position:].startswith(entry)):
            print("MATCH!", (current_position, end_position), entry, dic[entry])

